File: haskell/pe015.hs
module Main where
import Data.Vector (fromList, (!))

ways :: Int -> Int -> Int
ways m n = calc m n where
           calc 0 _ = 1
           calc _ 0 = 1
           calc x y = (c ! (x-1) ! y) + (c ! x ! (y-1))
           c = fromList [fromList [calc x y | y <- [0..n]] | x <- [0..m]]

main :: IO ()
main = print $ ways 20 20

File: python/pe015.py
cache = {}
def ways(r, c):
    if r == 0 or c == 0:
        return 1
    elif (r, c) not in cache:
        cache[(r, c)] = ways(r, c-1) + ways(r-1, c)
    return cache[(r, c)]
print(ways(20, 20))